# Cowpea adaptation to high temperatures in Africa
 
Detect candidate genes for adaptation to high temperature in African cowpea using via environmental genome wide association study (envGWAS).

* The current manuscript is available as: \
Akakpo R, Lee EJ, Pacheco JB, Rios EF, Kantar MB, Ousmane B, Volz KM, Akinmade H, Getino L, Boote KJ, et al. 2025. Identification of genetic variation associated with high-temperature tolerance in cowpea. bioRxiv. doi:10.1101/2025.04.09.647789

## UCR collaborators cowpea GWAS
[cowpea_gwas](https://github.com/cjfiscus/cowpea_gwas)

## Description
Using SNP data available from IITA Core accessions, structure of population analysis in ~600 accessions of African cowpea will be performed. Various population Genetic analysis including population-based estimate of recombination rate  [Ped/pop method](doi:10.1101/gr.067181.107). Using the same data we will carrie out environmental genome wide association study . 

## Genotyping datasets
- Genotyping dataset uses UCR Minicore Cluster file generated by Muñoz-Amatriaín M [SNP genotyping Matrix](https://www.dropbox.com/s/qiw024smgmj7snd/IITA_Core_SNPs_MinicoreCluster.xlsx?dl=0) 

## SNP position
- SNP positions from the iSelect design file are listed Github [SNP_Utils](https://github.com/MorrellLAB/cowpea_annotation/tree/main/Results)

## Passport data
- A Full passport data includes market collections & breeding material was provided by IITA (~1400 accessions) [Full passport data](https://www.dropbox.com/s/33pkna1l2ckjrgp/IITA_Core_2021_PassportData.xlsx?dl=0)
- Among them, some samples have localities, but no latitude and longitude.  We curate the data for a cleaned passport data file that does not include market collection or breeding material.  Cheking for matching information between country, Latitude/Longitude and location information will be done. Maybe possible to supplement with inferred latitude & longitude.
- Cleaned passport data will be upload on GitHub [Cleaned passport data](...) 

## Bioclimatic data source
- WorldClim is environmental data worldwide. We will focus on mean precipitation and temperature data, but also on growing month data [WorldClim](https://www.worldclim.org)

## Analysis Steps and Tools

### EnvGWAS

1. Performing envGWAS for BIO 1-19
2. Performing envGWAS for Monthly Rainfall, Max temperature and Min temperature

###  Population structure and differentiation 

1. Run structure analysis using [STRUCTURE](https://web.stanford.edu/group/pritchardlab/structure.html).  Try the analysis for correlated & uncorrelated allele frequency i.e. models with admixture & no admixture
As structure is time consuming for high dataset, get use LD pruning in [Plink](https://zzz.bwh.harvard.edu/plink/) to get a subset of the total SNPs to run structure analysis. Introduce iterations in the parameters file, [important](https://rosenberglab.stanford.edu/clumpp.html) for running structure

2. Run Procrustes analysis of PCA [Procruste PCA](PLoS Genet. 8: e1002886)

3. Fst Population differentiation test
 
Estimate FST for each SNP between subgroups (based on structure subgroups and/or geographic groups) and Identify outliers SNPs candidate for adaptive selection.

### Gene association tests

1. Identify the gene underling candidates SNPs using Bedtools intersect [Bedtools](https://bedtools.readthedocs.io/en/latest/), Snp2Gene Python program and appropriate [GFF annotation files](https://github.com/MorrellLAB/cowpea_annotation/blob/main/Results/IT97K-499-35_v1.0/iSelect_genotyped.vcf)
2. Perform Gene Ontology Tests to identify gene functions using the standard Fisher’s exact tests implemented in the R package [TopGO](https://doi.org/10.1093/bioinformatics/btl140)

3. LD of "hits" from iSelect to resequencing can use 36 cowpea [SNP calls](https://github.com/MorrellLAB/Env_Assoc/tree/master/script/LD_Analysis)

### Population Genetic analysis

1. Run Mosdepth for "callable regions" 
2. Calculate average per site heterozygosity (pi) using [Pixy](https://pixy.readthedocs.io/en/latest/)
3. Do Population-based estimate of recombination rate [Ped/pop method](doi:10.1101/gr.067181.107)
4. Infer population history using [SMC++ methods](https://doi.org/10.1038/ng.3748)

## Maps of collection localities

* Collection locality & temperature
[Maps](https://www.dropbox.com/sh/6wkzny0nhugh472/AAD6MSjfgUfeNMcKa_o26sAAa?dl=0)


## Repository Organization

### 00_format

The directory contains scripts for formating data.

1. Extract_Monthly_WorlClim_data.R is extracts mean monthly data from raster wordclim data;
2. Extract_Growing_Period_Clim_data.R is used formats genotype and phenotype text file into plink format (.tped, .tfam);
3. ped2binary.sh converts plink text files into binary and produces set of plink files for envgwas

### 01_envGWAS

This directory contains scripts to run envGWAS.

1. rel_mtx.sh calculates centered relatedness matrices for envgwas
2. mlm.sh is used to fit a mixed-linear models with [gemma](https://doi.org/10.1038/ng.2310)
3. prep_gemma_output_for_manathan_plot.R is an R script used to transform gemma output files into a suitable files for manhattan plot
4. manhattan_plot.R is an R script used to produce manhattan plot


