### Create a VCF to check SNP orientation
  
  * SNPs assays designed for the Illumina iSelect genotyping platform can be oriented either on the forward or reverse DNA strand relative to a genome assembly (or reference genome). Genotyping from the assay thus has SNPs in both orientations. To compare the SNPs to the reference genome they need to be flipped into the same orientation as the reference
  *  For cowpea, we use the ITK-499-35 assembled identified by Phytozome as "Vunguiculata_540_v1.0.fa." The iSelect 50K design sequence is available in supplemental materials for [Muñoz-Amatriaín et al. 2017](https://doi.org/10.1111/tpj.13404). The orientation of the SNP assays relative to the reference was reported in [Liang et al. 2023](https://doi.org/10.1002/tpg2.20319), and is available in the [cowpea_annotation](https://github.com/MorrellLAB/cowpea_annotation/tree/main/Results/IT97K-499-35_v1.0) GitHub repository. Positions were determined by mapping the [contextual sequence in FASTA format](https://github.com/MorrellLAB/cowpea_annotation/blob/main/iSelect_all.fas) with [BWA MEM](https:/doi.org/10.1093/bioinformatics/btp324) and then annotated with [SNP_Utils](https://github.com/MorrellLAB/SNP_Utils). The [SAM file](https://github.com/MorrellLAB/cowpea_annotation/blob/main/iSelect_cowpea_BWA.sam.gz) created by BWA MEM includes a flag value for each SNP, with a flag of 16 indicating the read (or contextual sequence was reverse complemented to map to the reference genome. A flag value of 2048 indicates a supplementary alignment, so 2048 + 16 = 2064 are supplementary alignments with reverse complement sequence. Thus iSelect SNPS where the reverse complement with a flag of 2064 don't appear among our genotyped variants. We created a list of all SNPs with a flag of 16, then use the Plink 1.9 [flip](https://www.cog-genomics.org/plink/1.9/data#flip) option to put all the reverse complemented variants onto the correct strand. 
    * With all the variants on the same strand, export to VCF demonstrates that many variants are not identifed with the same reference and alternate allele as in the VCF. To address this we tried a few approaches but ended up using Plink 2.0 [ref-allele](https://samtools.github.io/bcftools/howtos/plugin.fill-tags.html) option with the VCF of SNP positions noted above. This seems to work! Information on alternate and minor allele frequency is absent from VCF files created with Plink but is easy to add with [`bcftools +fill-tags`](https://samtools.github.io/bcftools/howtos/plugin.fill-tags.html).
    
   
